Introduce `ExperimentRestClient` and basic domain models

wzieba opened the pull request:
### Description

This PR is a first one of a few that aim to remove `FluxC` as a dependency for `experimantation` project.

In this PR I introduce `ExperimentRestClient` with logic of parsing fetched JSON and mapping it to domain models.

I decided to not introduce a feature branch to keep things simpler: newly introduced classes are not visible to consumers (except domain models, which are useless on its own) so it shouldn't cause confusion.

The logic is based on FluxC implementation, especially `org.wordpress.android.fluxc.network.rest.wpcom.experiments.ExperimentRestClient`

### How to test?

At this moment, testing is not required or possible due to lack of an entry point. The testing is covered with unit tests, and in next PRs I'll prepare some test scenarios.

wzieba committed:
Introduce necessary dependencies

`okhttp` and `moshi` are used by possible consumers. Code generation is faster and safer in the runtime, hence `moshi-codegen`

wzieba committed:
Introduce basic Assignments fetch with mapping to domain layer

wzieba committed:
Extract url builder to a class

So we can use a double in tests

wzieba committed:
Make domain classes `data`

Usually, using `data class` in a public contract is not the best idea, but I think it's fine with this internal-only library. We don't have to be *that* strict with backward compatibility

wzieba committed:
Introduce `Clock` interface and `SystemClock`

So it can be replaced with a double in tests

wzieba committed:
Introduce unit tests for fetching variations from remote

wzieba committed:
Adjust parsing logic to cover invalid JSON case

wzieba committed:
Merge branch 'bump_ktlint_to_0_50_0' into experimentation_without_fluxc_remote_layer

wzieba committed:
Fix `ktlintFormat` crash

https://github.com/pinterest/ktlint/issues/1195#issuecomment-1009027802

wzieba committed:
Address ktlint formatting rules

wzieba committed:
Remove unused `kotlinxCoroutinesExperimentationVersion` extension

wzieba committed:
Fix `ksp` plugin declaration

To match to the current project's convention

wzieba committed:
Make `UrlBuilder` internal

It doesn't have to be exposed to library consumer

ParaskP7 commented on experimentation/build.gradle:
```diff
@@ -36,20 +37,30 @@ android {
 
 dependencies {
+    def okHttp = "4.12.0"
+    implementation "com.squareup.okhttp3:okhttp:$okHttp"
```
Minor (🔍): FYI: There exist an advise from `Dependency Analysis` on transitive dependencies related to `okio`:

```
These transitive dependencies should be declared directly:
  implementation 'com.squareup.okio:okio:3.7.0'
```

wzieba replied:
I wonder why though - we don't use any code from `okio` directly 🤔

ParaskP7 replied:
Yea, I was wondering about that too, maybe the generate code by `Moshi` has to do with that, not sure, didn't look at the dependency graph, just searched for import and me too couldn't finding anything `Okio` related? 🤔 

FYI: We can suppress/exclude that if need be.

ParaskP7 commented on experimentation/build.gradle:
```diff
@@ -36,20 +37,30 @@ android {
     testImplementation "org.mockito:mockito-inline:$mockitoInlineVersion"
+
+    testImplementation "com.squareup.okhttp3:mockwebserver:$okHttp"
+    testImplementation "org.jetbrains.kotlin:kotlin-test:${gradle.ext.kotlinVersion}"
```
Minor (🔍): FYI: There exist an advise from `Dependency Analysis` on transitive dependencies related to `kotlin-test`:

```
Unused dependencies which should be removed:
  testImplementation('org.jetbrains.kotlin:kotlin-test:1.9.24') { capabilities {
    requireCapability('org.jetbrains.kotlin:kotlin-test-framework-junit')
}}
```

wzieba replied:
Nice, TIL about `capabilities` 😮

wzieba replied:
Anyway, I removed this library because I used it only for assertions and we can use JUnit for this.

ParaskP7 replied:
Makes sense to me, thanks! 👍

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/remote/AssignmentsDtoMapper.kt:
```diff
@@ -0,0 +1,20 @@
+import com.automattic.android.experimentation.domain.Assignments
+import com.automattic.android.experimentation.domain.Variation
+
+internal object AssignmentsDtoMapper {
```
Suggestion (💡): Consider adding a test for the mapper, just because of the `control` variation and it being mapped from `null` or `control`.

wzieba replied:
Done 👍 41b49391b3ce1fd87b06f4d1c7629b4056ad40c5

ParaskP7 replied:
Awesome, the test LGTM, thanks! ❤️

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/remote/AssignmentsDtoMapper.kt:
```diff
@@ -0,0 +1,20 @@
+internal object AssignmentsDtoMapper {
+    fun AssignmentsDto.toAssignments(fetchedAt: Long): Assignments {
+        return Assignments(
+            variations = variations.mapValues { (key, value) ->
```
Minor (🔍): Destructured parameter `key` is never used, consider renaming it to `_`.

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/remote/ExperimentRestClient.kt:
```diff
@@ -0,0 +1,47 @@
+        return withContext(Dispatchers.IO) {
+            okHttpClient.newCall(request).execute().use { response ->
+                if (!response.isSuccessful) {
+                    Result.failure<Assignments>(IOException("Unexpected code $response"))
```
Minor (🔍): Consider remove the explicit `<Assignments>` type arguments.

ParaskP7 commented on experimentation/src/main/java/com/automattic/android/experimentation/remote/ExperimentRestClient.kt:
```diff
@@ -0,0 +1,47 @@
+                    Result.failure<Assignments>(IOException("Unexpected code $response"))
+                } else {
+                    runCatching {
+                        val dto = jsonAdapter.fromJson(response.body!!.source())!!
```
Question (❓): You prefer using `runCatching` with `!!` over explicitly returning a success/failure result, right? Is there a way for us to get rid of `!!` with `runCatching` or you don't mind as anything will flow through the default failure when that happens anyway? 🤔

wzieba replied:
I think `runCatching` is a good fit because it returns the `Result` type - just what we need here. It protects against the NPE (from `!!`) but also from an incorrect JSON we parse here.

> Is there a way for us to get rid of !! with runCatching or you don't mind as anything will flow through the default failure when that happens anyway? 🤔

I don't mind much here about it 😅 I think it's okay to wrap here all exceptions - I understand it's usually an anti-pattern, but we don't want ever to have uncaught exception in `fetchAssignments`

ParaskP7 replied:
Let's then move forward with this solution! 👍 

> ...but also from an incorrect JSON we parse here.

💯

ParaskP7 commented on experimentation/src/test/java/com/automattic/android/experimentation/remote/ExperimentRestClientTest.kt:
```diff
@@ -0,0 +1,82 @@
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+internal class ExperimentRestClientTest {
```
Praise (❤️): Good job introducing this "integration" test with `MockWebServer` here!

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed this PR as per the instructions, everything LGTM, good job! 🌟 

FYI: I also traversed the [FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android) related repo and its [ExperimentRestClient](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/67b716889eb864763d703eb6d8c7519f73ad3157/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/experiments/ExperimentRestClient.kt) + [AssignmentsModel](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/67b716889eb864763d703eb6d8c7519f73ad3157/fluxc/src/main/java/org/wordpress/android/fluxc/model/experiments/AssignmentsModel.kt) implementation and everything seems to be mapped 1-to-1 with your `Moshi` solution here, well done. 🥇 

> I decided to not introduce a feature branch to keep things simpler: newly introduced classes are not visible to consumers (except domain models, which are useless on its own) so it shouldn't cause confusion.

👍 

> At this moment, testing is not required or possible due to lack of an entry point. The testing is covered with unit tests, and in next PRs I'll prepare some test scenarios.

👍 

-----

I have left one question (❓), a suggestion (💡) and some minor (🔍) comments for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

wzieba committed:
Clean up from unnecessary syntax elements

wzieba committed:
Add unit test to verify dto to domain mapping

wzieba committed:
Drop `kotlin-test` dependency

We were only using this for a few assertions. But we can use assertion method from JUnit, which are just fine for the use cases here.

wzieba committed:
Ignore warning on `transitive dependencies should be declared directly` for `okio`

wzieba committed:
Merge branch 'trunk' into experimentation_without_fluxc_remote_layer

wzieba commented:
Thank you for the review @ParaskP7 ! I've applied your suggestions, see you in next PRs 🙌

wzieba committed:
Address ktlint code analysis
